helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java

```diff
@@ -312,6 +312,17 @@ private void forceRebalance(HelixManager manager, ClusterEventType eventType) {
     changeContext.setType(NotificationContext.Type.CALLBACK);
     String uid = UUID.randomUUID().toString().substring(0, 8);
     ClusterEvent event = new ClusterEvent(_clusterName, eventType, uid);
+
+    Optional<String> leaderSession = manager.getSessionIdIfLead();
+    // If session is not present, this cluster manager is not leader for the cluster.
+    if (!leaderSession.isPresent()) {
+      logger.warn("Cluster manager {} is not leader for {}. Event {} is discarded.",
+          manager.getInstanceName(), manager.getClusterName(), event);
+      return;
+    }
+
+    // Pipeline should be run and Zk writes should be completed by the event session.
+    event.addAttribute(AttributeName.EVENT_SESSION.name(), leaderSession.get());
```

Comment:
In HandleEvent you already check if the sessionId matches the expected sessionId, why do we check again here?

Reply:
The purpose of this check here is to only add events to the queue if and only if controller is leader. We attach the session id as event session id. If controller is not leader, we discard the event and don't push it to queue.

In HandleEvent(), we check it as a filter if we should start the pipeline for the event. Because during the event in the queue, controller's session may change. If it changes, there is no need to start the pipeline for the stale event.

Reply:
Then when we created the event, just add the current session to the event attributes, is that good enough? We will check whether current controller is a leader and whether session matches anyway in handleEvent?

Reply:
What I meant, just need one line in the event create places:

    event.addAttribute(AttributeName.EVENT_SESSION.name(), manager.getSessionId());

Reply:
> Then when we created the event, just add the current session to the event attributes, is that good enough? We will check whether current controller is a leader and whether session matches anyway in handleEvent?

It is OK to check session when handleEvent, which could reduce the checks when pushing event. Downside is when we create the events, manager session has already expired. Then we still push stale events to the event queue.

Reply:
Since there are 3 places creating new events, I've moved the check to handleEvent().